File: ui-bookmarks/src/main/java/l/files/ui/bookmarks/menus/BookmarkMenu.java
package l.files.ui.bookmarks.menus;

import android.view.Menu;
import android.view.MenuItem;
import l.files.bookmarks.BookmarksManager;
import l.files.fs.Path;
import l.files.ui.base.app.OptionsMenuAction;
import l.files.ui.bookmarks.R;

import static android.view.Menu.NONE;
import static l.files.base.Objects.requireNonNull;

public final class BookmarkMenu extends OptionsMenuAction {

    private final BookmarksManager bookmarks;
    private final Path path;

    public BookmarkMenu(Path path, BookmarksManager bookmarks) {
        super(R.id.bookmark);
        this.bookmarks = requireNonNull(bookmarks, "bookmarks");
        this.path = requireNonNull(path, "path");
    }

    @Override
    public void onCreateOptionsMenu(Menu menu) {
        super.onCreateOptionsMenu(menu);
        menu.add(NONE, id(), NONE, R.string.bookmark).setCheckable(true);
    }

    @Override
    public void onPrepareOptionsMenu(Menu menu) {
        super.onPrepareOptionsMenu(menu);
        MenuItem item = menu.findItem(id());
        if (item != null) {
            item.setChecked(bookmarks.contains(path));
        }
    }

    @Override
    protected void onItemSelected(MenuItem item) {
        boolean checked = item.isChecked();
        item.setChecked(!checked);
        if (checked) {
            bookmarks.remove(path);
        } else {
            bookmarks.add(path);
        }
    }

}

File: ui-bookmarks/src/main/java/l/files/ui/bookmarks/menus/BookmarkMenu.kt
package l.files.ui.bookmarks.menus

import android.view.Menu
import android.view.Menu.NONE
import android.view.MenuItem
import l.files.base.lifecycle.SetLiveData
import l.files.ui.base.app.OptionsMenuAction
import l.files.ui.bookmarks.R
import java.nio.file.Path

class BookmarkMenu(
  private val path: Path,
  private val bookmarks: SetLiveData<Path>
) : OptionsMenuAction(R.id.bookmark) {

  override fun onCreateOptionsMenu(menu: Menu) {
    super.onCreateOptionsMenu(menu)
    menu.add(NONE, id(), NONE, R.string.bookmark).isCheckable = true
  }

  override fun onPrepareOptionsMenu(menu: Menu) {
    super.onPrepareOptionsMenu(menu)
    menu.findItem(id())?.isChecked = bookmarks.contains(path)
  }

  public override fun onItemSelected(item: MenuItem) {
    val checked = item.isChecked
    item.isChecked = !checked
    if (checked) {
      bookmarks.remove(path)
    } else {
      bookmarks.add(path)
    }
  }
}
